src/Uno.UI/Controls/Window.macOS.cs

```diff
@@ -27,12 +27,92 @@ namespace Uno.UI.Controls
 	/// </summary>
 	public partial class Window : NSWindow
 	{
+		private static readonly WeakAttachedDictionary<NSView, string> _attachedProperties = new WeakAttachedDictionary<NSView, string>();
+		private const string NeedsKeyboardAttachedPropertyKey = "NeedsKeyboard";
+
+		private InputPane _inputPane;
+		private WeakReference<NSScrollView> _scrollViewModifiedForKeyboard;
+
 		/// <summary>
 		/// ctor.
 		/// </summary>
 		public Window(CGRect contentRect, NSWindowStyle aStyle, NSBackingStore bufferingType, bool deferCreation)
 			: base(contentRect, aStyle, bufferingType, deferCreation)
 		{
+			_inputPane = InputPane.GetForCurrentView();
+			_inputPane.Window = this;
+		}
+
+		public static void SetNeedsKeyboard(NSView view, bool needsKeyboard)
+		{
+			if (view != null)
+			{
+				_attachedProperties.SetValue(view, NeedsKeyboardAttachedPropertyKey, (bool?)needsKeyboard);
+			}
+		}
+
+		private static bool GetNeedsKeyboard(NSView view)
+		{
+			var superViews = view.FindSuperviews().ToList();
+			superViews.Insert(0, view);
+
+			return superViews.Any(superView => _attachedProperties.GetValue(superView, NeedsKeyboardAttachedPropertyKey, () => default(bool?)).GetValueOrDefault());
 		}
+
+		private static bool NeedsKeyboard(NSView view)
+		{
+			return view is NSTextView
+				|| view is NSTextField
+				|| GetNeedsKeyboard(view);
+		}
+		
+		public BringIntoViewMode? FocusedViewBringIntoViewOnKeyboardOpensMode { get; set; }
+
+
+		internal void MakeVisible(NSView view, BringIntoViewMode? bringIntoViewMode, bool useForcedAnimation = false)
+		{
+			if (view == null)
+			{
+				return;
+			}
+
+			if (bringIntoViewMode == null)
+			{
+				return;
+			}
+
+			var scrollView = view.FindSuperviewsOfType<NSScrollView>().LastOrDefault();
+			_scrollViewModifiedForKeyboard = new WeakReference<NSScrollView>(scrollView);
+
+			if (scrollView == null)
+			{
+				this.Log().Warn("Keyboard will show, but we cannot find any ScrollView with enough space for the currently focused view, so it's impossible to ensure that it's visible.");
+				return;
+			}
+
+
+			var scrollViewRectInWindow = scrollView.ConvertRectFromView(scrollView.Bounds, scrollView);
+
+			var keyboardTop = (nfloat)_inputPane.OccludedRect.Top;
+
+			var keyboardOverlap = scrollViewRectInWindow.Bottom - keyboardTop;
+			if (keyboardOverlap > 0)
+			{
+				scrollView.ContentInsets = new NSEdgeInsets(0, 0, keyboardOverlap, 0);
+			}
+
+			var viewRectInScrollView = CGRect.Empty;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this useless assignment to local variable 'viewRectInScrollView'.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5303491)